### ISSUE TITLE
Tpetra: Make CrsMatrix MatrixMarket I/O memory scalable

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#353]
@trilinos/tpetra
Epic: #769.

MatrixMarket input and output (I/O) for Tpetra::CrsMatrix currently gathers the whole matrix to Process 0 before writing it.  This is not memory scalable.  "Memory scalable" means that no single process has to store more than a small constant factor times the maximum per-process memory usage of the distributed data structure.  See also #352 and #1017.

Tasks: 
  - [x] #1018: Write functions for packing and unpacking matrix triples (i, j, A(i,j)).
  - [x] #1025: Implement DistObject subclass that communicates by triples, using the above functions.
  - [ ] Rewrite CrsMatrix reader in the following way:
    - [x] #1031: Deal out chunks of triples to all the processes, in card-dealer fashion.
    - [ ] Push all the triples into the aforementioned DistObject subclass.
    - [ ] Do an Export of the subclass to the desired 1-to-1 Map.  (If we wanted to be fancy, we could let CrsMatrix be the direct target of that Export.)
    - [ ] As the Matrix Market reader already does, redistribute again if the original desired row Map is not 1-to-1.

**Comment by mhoemmen** [comment]
I include both CrsMatrix and CrsGraph in this issue, because they use nearly the same file format and share a lot of I/O code.  Note that this depends on #345.

**Comment by aprokop** [comment]
Also, some of the included headers are probably not used. Could include-what-you-use.org tool help with that for Trilinos in general?

**Comment by mhoemmen** [comment]
I haven't found that tool really useful for Tpetra.  I feel better removing them from Tpetra Matrix Market I/O than I do removing them from Teuchos_ConfigDefs.hpp (see the relevant issue), though.

**Comment by mhoemmen** [comment]
I just split off CrsGraph I/O into its own issue, #1017.

**Comment by mhoemmen** [comment]
I closed this issue by mistake.  Wish it were done ;-)

@jjellio , @tjfulle , this issue is relevant to current benchmarking efforts.

**Comment by tjfulle** [comment]
For some of the benchmarking problems, matrix I/O is the longest running part of the problem - any speed up would be awesome!

**Comment by mhoemmen** [comment]
@tjfulle :-) I'm pretty close to getting the DistObject subclass done.  I don't have to be fancy and worry about threads for now -- the point is MPI process memory scalability.  Also, it distributes entries by row, just like CrsMatrix, so it's not ideal load balancing.  However, it should be good enough for now.

**Comment by kddevin** [comment]
#7045 begins to address this issue
